src/services/fileStorage/strategies/awsS3.js

```diff
@@ -1,46 +1,105 @@
+		signatureVersion: 'v4',
+		s3ForcePathStyle: true,
+		sslEnabled: true,
+		accessKeyId: provider.accessKeyId,
+		secretAccessKey: provider.secretAccessKey,
+		region: provider.region || 'eu-de',
+		endpointUrl: provider.endpointUrl,
+		cors_rules: getCorsRules(),
+	});
+	awsConfig.endpoint = new aws.Endpoint(provider.endpointUrl);
+	return awsConfig;
+};
+
+const chooseProvider = async (schoolId) => {
+	const provider = await storageProviderModel.aggregate([
+		{ $match: { isShared: true } },
+		{ $addFields: { freeBuckets: { $subtract: ['$maxBuckets', { $size: '$schools' }] } } },
+		{ $sort: { freeBuckets: -1 } },
+		{ $limit: 1 },
+	]).exec();
+	if (!Array.isArray(provider) || provider.length === 0) throw new Error('No available provider found.');
+	await storageProviderModel.findByIdAndUpdate(provider[0]._id,
+		{ $push: { schools: schoolId } },
+		{ safe: true, upsert: true }).lean().exec();
+	return provider;
+};
 
+const FEATURE_MULTIPLE_S3_PROVIDERS_ENABLED = Configuration.get('FEATURE_MULTIPLE_S3_PROVIDERS_ENABLED');
+// begin legacy
 let awsConfig = {};
-try {
-	//	awsConfig = require(`../../../../config/secrets.${prodMode ? 'js' : 'json'}`).aws;
-	/* eslint-disable global-require, no-unused-expressions */
-	(['production'].includes(process.env.NODE_ENV))
-		? awsConfig = require('../../../../config/secrets.js').aws
-		: awsConfig = require('../../../../config/secrets.json').aws;
-	/* eslint-enable global-require, no-unused-expressions */
-} catch (e) {
-	logger.warning('The AWS config couldn\'t be read');
+if (!FEATURE_MULTIPLE_S3_PROVIDERS_ENABLED) {
+	try {
+		//	awsConfig = require(`../../../../config/secrets.${prodMode ? 'js' : 'json'}`).aws;
+		/* eslint-disable global-require, no-unused-expressions */
+		(['production'].includes(process.env.NODE_ENV))
+			? awsConfig = require('../../../../config/secrets.js').aws
+			: awsConfig = require('../../../../config/secrets.json').aws;
+		/* eslint-enable global-require, no-unused-expressions */
+	} catch (e) {
+		logger.warning('The AWS config couldn\'t be read');
+	}
 }
+// end legacy
+
+const createAWSObject = async (schoolId) => {
+	const school = await schoolModel.findOne({ _id: schoolId }).lean().exec();
+	if (school === null) throw new NotFound('School not found.');
+
+	if (FEATURE_MULTIPLE_S3_PROVIDERS_ENABLED) {
+		const S3_KEY = Configuration.get('S3_KEY');
+		let provider = await storageProviderModel.findOne({ schools: schoolId }).lean().exec();
+		if (provider === null) provider = await chooseProvider(schoolId);
+		provider.secretAccessKey = CryptoJS.AES.decrypt(provider.secretAccessKey, S3_KEY)
+			.toString(CryptoJS.enc.Utf8);
+
+		return {
+			s3: new aws.S3(getConfig(provider)),
+			bucket: `bucket-${schoolId}`,
+		};
+	}
 
-const AbstractFileStorageStrategy = require('./interface.js');
-
-const createAWSObject = (schoolId) => {
+	// begin legacy
 	if (!awsConfig.endpointUrl) throw new Error('AWS integration is not configured on the server');
```

Comment:
```suggestion
	if (!awsConfig.endpointUrl) throw new Error('S3 integration is not configured on the server');
```
should we use S3 instead of any aws?